### ISSUE TITLE
@file is confused by symlinks (regression since 1.9.1)

### ISSUE THREAD

**Issue by jwakely** [doxygen/doxygen#9211]
**Describe the bug**
Doxygen gives an error when an input file is a symlink to a file of a different name, and a `@file` command refers to the symlink's name, not the target file.

**Expected behavior**
The symlinked file should be processed and its documentation added to the output

**Screenshots**

This is the output for running doxygen 1.9.3 against the attached tarball:
![Screenshot2](https://user-images.githubusercontent.com/1254480/158571020-d61a0384-cf22-4240-a5d9-a1c38993ab7e.png)

Note that `bar.h` is not a link and does not have any brief description.

For comparison, here is the correct output for doxygen 1.9.1 on the same files:

![Screenshot1](https://user-images.githubusercontent.com/1254480/158570755-f2d1d2c1-c196-4429-af16-d8b0e811b601.png)

`bar.h` is documented correctly.

**To Reproduce**

[symlinkbug.tar.gz](https://github.com/doxygen/doxygen/files/8261273/symlinkbug.tar.gz)

Untar this file and run `doxygen`

The output will include:

```
$ doxygen 
Doxygen version used: 1.9.3 (4bf942faa3128099a4fbc47162e11b0545487fd5)
Searching for include files...
Searching for example files...
Searching for images...
Searching for dot files...
Searching for msc files...
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Searching for files in directory /tmp/dox2/include
Reading and parsing tag files
Parsing files
Preprocessing /tmp/dox2/bar.h...
Parsing file /tmp/dox2/bar.h...
Preprocessing /tmp/dox2/include/baz.h...
Parsing file /tmp/dox2/include/baz.h...
...
Generating file sources...
Generating code for file bar.h...
error: file '/tmp/dox2/bar.h' not found
Generating code for file include/baz.h...
Generating file documentation...
Generating docs for file include/baz.h...
...
```

**Version**

The bug is present in `1.9.3 (4bf942faa3128099a4fbc47162e11b0545487fd5)` (self-compiled from Git HEAD yesterday).

The bug is **not** present in 1.9.1

**Comment by albert-github** [comment]
Small analysis:
The 1.9.3. (4bf942faa3128099a4fbc47162e11b0545487fd5) is a development version from August 19 2021, but not relevant for the analysis as with the current master `1.9.4 ()`

I see the error:
```
error: file '.../bar.h' not found
```
which already gives a bit of an indication.
Looking at the output I see in the 1.9.1 version:
![image](https://user-images.githubusercontent.com/5223533/158573415-018258b0-b830-4200-8880-e4b2e5a0c9c3.png)

and in the 1.9.3 version:
![image](https://user-images.githubusercontent.com/5223533/158573598-93ff0e3d-0342-417e-b911-44ccb2a40028.png)

we also see:
- that `bar.h`  is just  shown as text and without a brief description (following from the error message mentioned before.
- the level of `bar.h` did change to the place where `foo.h` resides (i.e. the symbolic link points to).
- the preprocessor (`doxygen -d preprocesssor`) output of `bar.h` gives:
   ```
   Preprocessing .../bar.h...
   Preprocessor output of .../bar.h (size: 0 bytes):
   ---------
   00001
   ---------
   ```

I think the problem is caused due to the new way doxygen parses the file system (with the `filesystem` include file) and that here the symbolic links are not handled properly here or incorrectly called by doxygen.

**Comment by jwakely** [comment]
Oops, sorry, yes, I have three versions sitting around. The bug is still present with `1.9.4 (1231a9eed2b5b4b74295d0fe400fb37203d8101b)` from HEAD yesterday.

**Comment by jwakely** [comment]
The bug is in `readDir` at line 10330:

```c++
        std::string name=cfi.fileName();
        std::string path=cfi.dirPath()+"/";
        std::string fullName=path+name;
```

This calls `dirPath()` which uses the default argument, so is equivalent to `dirPath(absPath=true)`, but that means we get the directory portion of the absolute canonicalized path, which has resolved the symlink, and then concatenate the file part of the symlink, but that's nonsense, there is no such file.

**Comment by jwakely** [comment]
Specifically, when `$PWD` is `/tmp/dox2` (as in the output I showed above), the path of the symlink is `/tmp/dox2/include/bar.h` so `cfi.fileName()` returns `"bar.h"`, and `cfi.dirPath(true)` returns `"/tmp/dox2"` so the concatenated result is `"/tmp/dox2/bar.h"` which doesn't exist. The symlink is `/tmp/dox2/include/bar.h` and the target is `/tmp/dox2/foo.h` but the concatenated string refers to neither.

**Comment by jwakely** [comment]
The fix is trivial:

```diff
--- a/src/doxygen.cpp
+++ b/src/doxygen.cpp
@@ -10328,7 +10328,7 @@ static void readDir(FileInfo *fi,
           )
       {
         std::string name=cfi.fileName();
-        std::string path=cfi.dirPath()+"/";
+        std::string path=cfi.dirPath(false)+"/";
         std::string fullName=path+name;
         if (fnMap)
         {
```

**Comment by jwakely** [comment]
But I don't understand what this is supposed to be doing. `cfi.path()` would already return that string, wouldn't it?

If the purpose is to get an absolute canonical path, use `cfi.absFilePath()` instead of this concatenation. Or if the purpose is to get an absolute **un-canonicalized** path (i.e. without resolving symlinks) then `dirPath()` can't be used, because it also canonicalizes the path when called with `absPath=true`.

`std::filesystem::absolute(cfi.filePath()).string()` would work for an absolute un-canonicalized path.

Maybe the correct fix is for `FileInfo::dirPath(true)` to not use `absFilePath()`, so it can get an absolute un-canonicalized path:

```diff
--- a/src/fileinfo.cpp
+++ b/src/fileinfo.cpp
@@ -139,13 +139,13 @@ std::string FileInfo::dirPath(bool absPath) const
   std::string result;
   if (absPath)
   {
-    result = absFilePath();
+    result = fs::absolute(m_name);
   }
   else
   {
     result = m_name;
-    correctPath(result);
   }
+  correctPath(result);
   size_t pos = result.rfind('/');
   if (pos==std::string::npos)
   {
```

Or maybe `FileInfo::absFilePath()` should take a boolean argument saying whether to canonicalize, or just make the path absolute.

Without knowing what this API is supposed to be doing, I have no idea what the right fix is.

The `FileInfo` abstraction seems to be doing more harm than good here, personally I would just use `std::filesystem` directly, because at least it has a clearly defined API (I don't see any comments explaining what `FileInfo` member functions are **supposed* to be doing).

**Comment by albert-github** [comment]
Looks OK, will you make a pull request for it?

**Comment by albert-github** [comment]
Yes I think @doxygen should have a look at it.

**Comment by jwakely** [comment]
Hmm, nearly every use of `FileInfo::dirPath()` looks like it has this problem:

```
$ git grep dirPath
src/configimpl.l:      diaPath=dp.dirPath(TRUE)+"/";
src/docbookvisitor.cpp:        FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/docbookvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/docbookvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/doxygen.cpp:        std::string path=cfi.dirPath()+"/";
src/doxygen.cpp:          std::string dirPath = fi.dirPath(true);
src/doxygen.cpp:          if (paths && !dirPath.empty())
src/doxygen.cpp:            paths->insert(dirPath);
src/doxygen.cpp:              std::unique_ptr<FileDef> fd { createFileDef(QCString(dirPath+"/"),QCString(name)) };
src/fileinfo.cpp:std::string FileInfo::dirPath(bool absPath) const
src/fileinfo.h:    std::string dirPath(bool absPath = true) const;
src/htmldocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/htmldocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/htmldocvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/latexdocvisitor.cpp:        FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/latexdocvisitor.cpp:        FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/latexdocvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/mandocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/mandocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/mandocvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/pre.l:      QCString absName = QCString(fi.dirPath(TRUE))+"/"+fileName;
src/pre.l:    QCString absName = QCString(fi.dirPath(TRUE))+"/"+incFileName;
src/pyscanner.l:  return findPackageScopeFromPath(yyscanner,fi.dirPath(true).c_str());
src/rtfdocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/rtfdocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/rtfdocvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/xmldocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/xmldocvisitor.cpp:         FileDef *fd = createFileDef( cfi.dirPath(), cfi.fileName() );
src/xmldocvisitor.cpp:        fd = createFileDef( cfi.dirPath(), cfi.fileName() );
```

Any time `cfi.dirPath()` and `cfi.fileName()` are used together, you get the absolute canonicalized directory and the un-canonicalized filename, and combining those two might get `ENOENT` or might read a completely different file from the one intended.

**Comment by jwakely** [comment]
Another possible fix is to just make `FileInfo::absFilePath()` do what the name suggests, i.e. make the path absolute. Don't use `fs::canonical` because that gives you a completely different path when symlinks are involved.

```diff
--- a/src/fileinfo.cpp
+++ b/src/fileinfo.cpp
@@ -100,17 +100,8 @@ void FileInfo::correctPath(std::string &s)
 
 std::string FileInfo::absFilePath() const
 {
-  std::string result;
   std::error_code ec;
-  fs::path path(m_name);
-  if (fs::exists(path,ec))
-  {
-    result = fs::canonical(path,ec).string();
-  }
-  else
-  {
-    result = (fs::current_path(ec) / m_name).string();
-  }
+  std::string result = fs::absolute(m_name,ec).string();
   correctPath(result);
   return result;
 }
```

But both this and the patch for `FileInfo::dirPath` result in:

```
43 out of 82 tests failed
```

Whereas the minimal fix for `readFile` that just changes it to `dirPath(false)` results in:

```
All tests passed!
```

**Comment by jwakely** [comment]
Qt's `QFileInfo` gets this right, with separate functions for [`absoluteFilePath`]( https://doc.qt.io/qt-5/qfileinfo.html#absoluteFilePath) and `canonicalFilePath`. The docs for the former even say:

> In contrast to [canonicalFilePath](https://doc.qt.io/qt-5/qfileinfo.html#canonicalFilePath)(), symbolic links or redundant "." or ".." elements are not necessarily removed.

And neither `QFileInfo::dir()` nor `QFileInfo::absoluteDir()` return a canonical path.

In contrast, doxygen's `FileInfo::dirPath(absPath=true)` returns a canonical path if the file exists (because that's what `absFilePath` does) and a non-canonical path if it doesn't exist, and `FileInfo::dirPath(false)` never canonicalizes the result.

Messy.

**Comment by dirk-thomas** [comment]
Please see #9294 for another incarnation of absolute vs. canonical problems when symlinks are involved.

**Comment by doxygen** [comment]
@jwakely The `FileInfo` class attempts to reimplement the very old version of QFileInfo I used before using STL functionality. So looks indeed the translation was not done correctly for these functions (i.e. works ok for regular files but breaks down for symlinks).

**Comment by doxygen** [comment]
@jwakely Please verify if the referenced commit fixes the problem.

**Comment by jwakely** [comment]
Yes, this seems to fix my issue, thanks.

**Comment by jwakely** [comment]
I have `FULL_PATH_NAMES=NO` and yet I'm seeing a complete directory hierarchy in the File List. A completely separate hierarchy from the one where I'm running doxygen, and the `INPUT`.

**Comment by albert-github** [comment]
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- Can you please also attach the images you see with the current version and the 1.9.1 version?

**Comment by jwakely** [comment]
Will do

**Comment by jwakely** [comment]
See above for screenshots of my real project.

Here's a self-contained reproducer:

[filelist-bug.tar.gz](https://github.com/doxygen/doxygen/files/8671911/filelist-bug.tar.gz)

Untar it, then `cd build && doxygen`

With 1.9.1 the file list looks like:

![191](https://user-images.githubusercontent.com/1254480/167903784-3744a1b8-9d51-44a3-bc44-c25bf838b51c.png)

But with 1.9.5 86fe8ce908bfd6782633d2d9647f2159c2f08a44 I get:

![195](https://user-images.githubusercontent.com/1254480/167903885-76c212b8-ba81-4f4d-9e40-f29c3c2e9f4a.png)

Is there a way to not get that directory tree? It seems like `FULL_PATH_NAMES=NO` is being ignored. The layout of the directories in the source tree is irrelevant to the end user and doesn't correspond the the layout of the files that will be installed. I just want the file names.

Alternatively, is there a way to not add the File List to the tree view at all? I want the Modules, Classes, Namespaces etc. but I don't actually need users to know about `<bits/internal_header.h>` and `<impl_details.h>`, only the public API headers that are documented with `@headerfile` commands.

**Comment by jwakely** [comment]
But at least that directory tree is accurate since 62bea90229c56e71d1fe150cb4b19d5e4fc6e9e0, before that it had complete nonsense:

![bad](https://user-images.githubusercontent.com/1254480/167908123-23f9a1a1-1ff1-4e25-add2-bc31944506cb.png)

There is no `real/y.h` anywhere in the project, there is a symlink `include/y.h` that points to `real/y.cc`, so at least that's fixed in current master.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
